### PR TITLE
Implements :default_locale and translation fallback chains

### PR DESCRIPTION
**Pull request by kipcole9**
This PR implements the proposal in [issue #73](https://github.com/crbelaus/trans/issues/73):

* Declare the default locale for a schema
* Implement fallback chains as an option to `Trans.Translator.translate/{2,3}`
* Support default locale and fallback chains in `Trans.QueryBuilder.translated/3`

During the development of this PR I realised that Postgres JSONB operators can return JSON `'null'` in one case and SQL `NULL` in all other cases. This is explained very well in [this blog post](http://mbork.pl/2020-02-15_PostgreSQL_and_null_values_in_jsonb).

Therefore these PR standardises on SQL `NULL` for all cases since I believe the consistency helps and the use of SQL `NULL` means returns of `nil` in Elixir which is probably more expected.

All original tests pass however its not clear (to me) if this is considered a breaking change.  I will finish updating documentation and then the PR would be considered final.

Comments, suggestions, improvements welcome.

### CODE REVIEW

**Comment by crbelaus** [comment]
Hey @kipcole9 
This is a wonderful pull request. Thank you for all the effort.

Unfortunately I won't be able to take a detailed look at the code for the next two weeks. But I will go back to this as soon as I can.

Thanks again!

**Comment by kipcole9** [comment]
@crbelaus thanks for the feedback and encouragement. Definitely no rush.

I have added a further commit that implements support for dynamic fallback chains. For various reasons explained in the commit message, that has necessitated the implementation of a database function as requirement. To make that easy there is a migration generator and the migration path is no more than:

* `mix deps.update trans`
* `mix trans_gen_translate_function`
* `mix ecto.migrate`

There are no changes required to application code. All existing tests pass with two tests potentially returning false positives due to the underlying code not returning the base column value when the translation for a locale does not exist. I have made annotations in the test file for discussion.

I appreciate that requiring a database function and therefore a migration to be generated and instantiated may not be in line with your goals. But it might make for an interesting discussion :-)

There are some good benefits to using the database function however:

1. It means we can now use `translated/3` and its new friend `translated_as/3` in the `SELECT` clause making it very straight forward to build translatable queries.
2. Depending on the query, it should not be necessary for all the translations to be sent over the wire for each query, saving bandwidth and reducing latency.
3. The translation itself is performed in the database, not on the Elixir side. Pros and cons of course, but in this case it may well be a good outcome. It needs profiling to be more certain.

**Comment by kipcole9** [comment]
This is the last "feature" commit that fulfils the principals from the original proposal. It adds a `translations/3` macro to be used in Ecto schema definitions.  Apart from the clean interface, it makes it easier to integrate with other libraries such as `ex_cldr`.  Here is one example:
```elixir
    defmodule Magazine do
      use Ecto.Schema
      use Trans, translates: [:title, :body], default_locale: :en

      schema "magazine" do
        field :title, :string
        field :body, :string
        translations :translations, Translations, MyApp.Cldr.known_locale_names()
      end
    end
```
In the example, `MyApp.Cldr.known_locale_names()` returns the configured locale names for this `ex_cldr` backend module.

**Comment by kipcole9** [comment]
@crbelaus, I'm going to go ahead an implement this as an `ex_cldr_trans` library for now so I can move forward onto some other areas that I want to complete before my talk at ElixirConf EU in June.  Of course if you decide later on to proceed with this I will very happily redirect my lib back to `trans` since the work is yours.

**Comment by kipcole9** [comment]
Oh, no pressure from me, take whatever time you need to review.  There's a lot going on so happy to discuss at your convenience. Its quite possible this is too much of a change overall, depending on your objectives.

File: lib/trans/query_builder.ex [hunk 1]
```diff
@@ -95,55 +112,115 @@ if Code.ensure_loaded?(Ecto.Adapters.SQL) do
...
+    # the binding.  If the binding ends in a digit then we assume
+    # this is actually the table alias.  If it is not, append a `0`
+    # and treat it as the table alias.  This because its not
+    # possible to know the table alias at compile time.
+    @digits Enum.map(0..9, &to_string/1)
+
+    defp table_alias({schema, _, _}) do
+      schema = to_string(schema)
+      if String.ends_with?(schema, @digits), do: schema, else: schema <> "0"
     end
```
Comment (crbelaus) [inline lib/trans/query_builder.ex:224]: This is really clever 🙇

**Review by crbelaus — APPROVED** [review]
This is an awesome contribution @kipcole9. I've learned a lot reviewing the changes 🙇 

I love the new `translations` macro. The only downside of structured translations vs. unstructured translations was that the structured translations require more ceremony and code. The new `translations` macro makes all this go away since the ceremony is automatically generated for us.

I also found the database function for translation fallbacks really nice and well done. 

Overall I think that his is a wonderful improvement for `trans` and I will release a new version with this changes. I think that all this new improvements are so great that I'm leaning towards publishing a new major version, so this would be `trans` 3.0.

What do you think? You have a lot of experience maintaining the [ex_cldr](https://hex.pm/packages/ex_cldr) family of libraries (and more) so I would love to hear your advice on this.

Thanks again 🙌

**Comment by kipcole9** [comment]
Glad the changes are ok with, you created a really solid library and its genuinely fun to work on the additions.

Given you are thinking to remove unstructured translations - combined with the addition of a database function and required migrations - it sounds reasonable to me to call this a major release.  Also because while all your original tests pass its possible there are edge cases that have slightly different behaviour (`NULL` returns versus `'null'`)

The documentation also needs some updating regarding the migration generator.   I'm fine to work on that and you remove the parts you planned to remove?

**Comment by crbelaus** [comment]
Sounds great!
